spring-cloud-kubernetes-fabric8-discovery/src/test/java/org/springframework/cloud/kubernetes/fabric8/discovery/Fabric8EndpointsAndEndpointSlicesTests.java

```diff
@@ -0,0 +1,338 @@
+				Set.of(namespace), true, 60, false, "", Set.of(), labels, "", null, 0, endpointSlices);
+		KubernetesCatalogWatch watch = new KubernetesCatalogWatch(mockClient(), properties, NAMESPACE_PROVIDER);
+
+		if (endpointSlices) {
+			watch = Mockito.spy(watch);
+			Mockito.doReturn(new Fabric8EndpointSliceV1CatalogWatch()).when(watch).stateGenerator();
+		}
+
+		watch.postConstruct();
+		watch.setApplicationEventPublisher(APPLICATION_EVENT_PUBLISHER);
+		return watch;
+
+	}
+
+	KubernetesCatalogWatch createWatcherInSpecificNamespacesWithLabels(Set<String> namespaces,
+			Map<String, String> labels, boolean endpointSlices) {
+
+		// all-namespaces = false
+		KubernetesDiscoveryProperties properties = new KubernetesDiscoveryProperties(true, false, namespaces, true, 60,
+				false, "", Set.of(), labels, "", null, 0, false);
+		KubernetesCatalogWatch watch = new KubernetesCatalogWatch(mockClient(), properties, NAMESPACE_PROVIDER);
+
+		if (endpointSlices) {
+			watch = Mockito.spy(watch);
+			Mockito.doReturn(new Fabric8EndpointSliceV1CatalogWatch()).when(watch).stateGenerator();
+		}
+
+		watch.setApplicationEventPublisher(APPLICATION_EVENT_PUBLISHER);
+		watch.postConstruct();
+		return watch;
+
+	}
+
+	void endpoints(String namespace, Map<String, String> labels, String podName) {
+
+		EndpointAddress endpointAddress = new EndpointAddressBuilder()
+				.withTargetRef(new ObjectReferenceBuilder().withName(podName).withNamespace(namespace).build()).build();
+
+		EndpointSubset endpointSubset = new EndpointSubsetBuilder().withAddresses(List.of(endpointAddress)).build();
+
+		Endpoints endpoints = new EndpointsBuilder()
+				.withMetadata(new ObjectMetaBuilder().withLabels(labels).withName("endpoints-" + podName).build())
+				.withSubsets(List.of(endpointSubset)).build();
+		mockClient().endpoints().inNamespace(namespace).create(endpoints);
+	}
+
+	static void endpointSlice(String namespace, Map<String, String> labels, String podName) {
+
+		Endpoint endpoint = new EndpointBuilder()
+				.withTargetRef(new ObjectReferenceBuilder().withName(podName).withNamespace(namespace).build()).build();
+
+		EndpointSlice slice = new EndpointSliceBuilder().withMetadata(new ObjectMetaBuilder().withNamespace(namespace)
+				.withName("slice-" + podName).withLabels(labels).build()).withEndpoints(endpoint).build();
+
+		mockClient().discovery().v1().endpointSlices().inNamespace(namespace).resource(slice).create();
+
+	}
+
+	static void invokeAndAssert(KubernetesCatalogWatch watch, List<EndpointNameAndNamespace> state) {
+		watch.catalogServicesWatch();
+
+		verify(APPLICATION_EVENT_PUBLISHER).publishEvent(HEARTBEAT_EVENT_ARGUMENT_CAPTOR.capture());
+
+		HeartbeatEvent event = HEARTBEAT_EVENT_ARGUMENT_CAPTOR.getValue();
+		assertThat(event.getValue()).isInstanceOf(List.class);
+
+		assertThat(event.getValue()).isEqualTo(state);
+	}
+
+	// work-around for : https://github.com/fabric8io/kubernetes-client/issues/4649
```

Comment:
kind of un-expected from the fabric8 test extension... If I find time I will issue a PR against their repo also.